### ISSUE TITLE
YouTube videos to promote the VS Code workflow

### ISSUE THREAD

**Issue by ndward** [ev3dev/vscode-ev3dev-browser#61]
With the fixing of the brickd problem I feel the VS Code/Stretch workflow is working well and is worthy of promotion. I'm not aware of any serious problems other than the incorrect handling of file names and folder names that contain spaces (still not fixed in the 24 May 2018 image) and the font being too large when one types the WiFi password, making it impossible to distinguish E from F or I from L (still not fixed in the 24 May 2018 image).
The ev3dev VS Code / Stretch project is a work in progress (and no doubt always will be) but that should not stop promotion of the workflow now that it works as well as it does. Reasons not to promote the workflow now would include major bugs and major upcoming changes, particularly breaking changes. Are we aware of any bugs? Are there any major upcoming changes? One enhancement that I am particularly keen to see is the introduction of functions to mirror the Move Steering and Move Tank blocks of EV3-G. As webmaster of ev3python.com I regularly get emails from people eager to see such functions. See a discussion here: https://sites.google.com/site/ev3python/learn_ev3_python/using-motors/ev3g-like-motor-functions
I have made two videos, the first to show how easy it is to write and run EV3dev Python scripts using VS Code with the EV3 extension, and the second to show how to set up the EV3 and the computer to enable the workflow.
I made a candidate first video (17 minutes) and published it to YouTube as unlisted. I deleted that one and replaced it with a new one where I no longer speak of the need to switch between configurations - I realised that no configuration is necessary to run non-EV3 files locally: https://youtu.be/cqtRqsI6xMc
The setup video is now also awaiting your comments at https://youtu.be/-ztvWQFagNA
I welcome your comments as to the advisability of publishing such videos now, and specific  comments on the videos I just made (errors, things that are not clear…).

**Comment by dlech** [comment]
> I'm not aware of any serious problems other than the incorrect handling of file names and folder names that contain spaces (still not fixed in the 24 May 2018 image) and the font being too large when one types the WiFi password, making it impossible to distinguish E from F or I from L (still not fixed in the 24 May 2018 image).

I don't see any open issues for these. Can you provide a link?



> Are we aware of any bugs?

Not any major ones.



> Are there any major upcoming changes?

There is nothing planned at the moment that I am aware of.


> I welcome your comments as to the advisability of publishing such videos now, and specific comments on the video I just made (errors, things that are not clear…).

I don't any reason to postpone such videos.

As for the video itself, I didn't see any noteworthy errors. I think starting around the 12 minute mark, you are perhaps getting a bit too much into the details of how to write a python script compare to the high level overview of the rest of the video. Also, the easy 3 step process of "write, save, press F5" actually only needs to be a 2-step process. Pressing F5 automatically saves open files, so that dosn't need to be a separate step.

**Comment by ndward** [comment]
In case you missed the updates that I made to my previous comment, let me say again that I deleted the first video and replaced it with a new one where I no longer speak of the need to switch between configurations - I realised that no configuration is necessary to run non-EV3 files locally: https://youtu.be/cqtRqsI6xMc (17 minutes)
The setup video is now also awaiting your comments at https://youtu.be/-ztvWQFagNA (36 minutes)

**Comment by dlech** [comment]
These videos are really superb. The new video touches on all of the important things you need to know.

One thing that jumped out at me though was the recommendation to use the "Windows defaults" for a network connection. We've actually had a number of people that have reported problems that ended up being because they did select this option when they should not have. This option only helps in very specific circumstances. Because of this, I don't think it is a good idea to make this recommendation for general use.

Some minor suggestions:
* The clip with the actual EV3 robot is loud and it is hard to hear you talking at the same time
* There are simpler ways to extract the zip file that don't result in nested folders with the same name

**Comment by ndward** [comment]
I have modified my 'setting up' video based on the suggestions made by @dlech and have published the modified version to https://youtu.be/TNXqizQTZhs (36 minutes). Recall that the introductory video at https://youtu.be/cqtRqsI6xMc (17 minutes) should be watched first.

have also modified my site ev3pthon.com to begin promoting this workflow which I think is far superior, especially for beginners, and which I think could easily be used in schools. A colleague in my old school, European School 3 Brussels, will soon begin teaching classes based on the new workflow and she will provide feedback...

Regarding the recommendation to use Window defaults, which is causing problems for some users, I note that the page http://www.ev3dev.org/docs/tutorials/connecting-to-the-internet-via-usb/ still instructs users to set the Windows defaults.

**Comment by dlech** [comment]
The "Windows defaults" button sets a static IP address and gateway that match Windows Internet Connection Sharing. If you are using Wi-Fi, these will almost certainly be wrong. Also, if you are using USB or Bluetooth and Windows Internet Connection sharing is not configured, it will still look like the EV3 has a connection because of the manually assigned address, but you won't be able to actually ping the EV3.

FYI, the work-in-progress docs for setting up networking on ev3dev-stretch are here: http://docs.ev3dev.org/en/ev3dev-stretch/getting-started/ev3.html#establishing-a-connection

**Comment by bittner** [comment]
@ndward Just some quick feedback on the videos:

1. You may want to try out `ev3devlight` by @laurensvalk (as [discussed on Gitter](https://gitter.im/ev3dev/chat?at=5b2d4cf67d3bca737a01ea7f)), which is said to reduce the Python interpreter startup time on the brick and seems to allow to [steer two motors in parallel](https://github.com/laurensvalk/ev3devlight-examples/blob/master/drivebase.py#L11-L17) (if I'm not mistaken).

1. As [commented on the Setting up video](https://www.youtube.com/watch?v=TNXqizQTZhs), using [ev3py-runlocal](https://pypi.org/project/ev3py-runlocal/) may make the developer experience more smooth with regards to local development, including code completion and maybe even debugging.

1. Do you think you can produce a super-fast, **super-short intro video** that gives a quick, fascinating insight into the main features and workflow of ev3dev development (using VS Code and the ev3dev-browser plugin)? You know, everyone has a short attention span nowadays. :roll_eyes:

**Comment by WasabiFan** [comment]
@bittner Could you help us understand what it is you're looking for that makes you prefer the aforementioned alternative library over `ev3dev-lang-python`? I want to see what can be done to consolidate things so we can keep the userbase on a single platform. Certainly, Micropython support is a big plus (it's the next thing I'm going to work on after https://github.com/ev3dev/ev3dev-lang-python/pull/411, and is partially complete already); we've also made various improvements to module architecture for Stretch which have hopefully made load times better. API-wise, it seems like `ev3devlight` is a subset of `ev3dev-lang-python` -- the overall API structure is shared and most of the utilities already exist in Jessie, with more full-featured versions coming with the Stretch release of the library.


This thread should probably continue elsewhere (new issue, Gitter, etc.) but I'd like to hear your thoughts.

**Comment by bittner** [comment]
@WasabiFan I'm only addressing hick-ups @ndward is mentioning in his videos. From the top of my head:

1. "It's not yet possible to run two motors in parallel"
1. "You have to wait a minute each time for the Python interpreter to load"
1. "The red underlines in the editor are because the (ev3dev) library is not known locally; but it will be known on the robot"

These are obviously usability issues, disturbing defects of the development experience, that need to be solved to keep the entrance level low. @dlech [mentioned on Gitter](https://gitter.im/ev3dev/chat?at=5b296543aeb8fa0c0729e722) that MicroPython may not be happening on ev3dev anytime soon. Hence, I thought that rather `ev3devlight` might demonstrate an user experience improvement, one that may give enhancements on the main project a boost. I'm not trying to fragment the user base.

**Comment by dlech** [comment]
FWIW, I don't think 3. is actually a problem. It just requires installing the ev3dev python library on your host computer to make it go away (and get code completion working).

**Comment by bittner** [comment]
>  It just requires installing the ev3dev python library on your host computer

As mentioned on Gitter, I tried it and it doesn't. Probably because the library hits code that requires LEGO hardware, and fails. With `ev3py-runlocal` this problem is intentionally solved.

**Comment by dlech** [comment]
:man_shrugging: it works for me for code completion. Actually running the code is different of course.

**Comment by WasabiFan** [comment]
> "It's not yet possible to run two motors in parallel"

It is in Stretch. We have `MoveTank`, `MoveSteering` and `MoveJoystick` classes depending on exactly what you're trying to do. Feedback on those would be appreciated.

> "You have to wait a minute each time for the Python interpreter to load"

Hopefully this has been improved by our restructuring and will improve further once Micropython support arrives. Certainly, anything approaching a minute is a problem.

> "The red underlines in the editor are because the (ev3dev) library is not known locally; but it will be known on the robot"

It should work via installing the library on your dev machine. If not, please open an issue on the ev3dev-lang-python repo with the steps you took and error you're seeing and we can help you out. I have some potential theories.

**Comment by ndward** [comment]
I'm delighted to learn about those three functions. Can you please tell me where to find the documentation? Thanks.

As for installing ev3dev.ev3 on my Windows PC, I'm not sure that's so easy.

**Comment by WasabiFan** [comment]
@ndward they will be published with our initial Stretch release of the Python library soon. Right now the code is on GitHub but not released as a package yet. We're working on the documentation here: https://github.com/ev3dev/ev3dev-lang-python/pull/411.

The Python library should be installable via pip or via source on Windows;  I'm not sure if we have been publishing to the pip service though. Probably not.
